websockets/test_client_server.py

```diff
@@ -442,18 +442,41 @@ def test_invalid_status_error_during_client_connect(self):
         self.assertEqual(exception.code, 403)
         self.stop_server()
 
-    @unittest.mock.patch('websockets.server.read_request')
-    def test_connection_error_during_opening_handshake(self, _read_request):
-        _read_request.side_effect = ConnectionError
+    def do_error_during_opening_handshake(self, exception_cls):
+        written = []
 
-        self.start_server()
+        class WriteCollectingProtocol(WebSocketServerProtocol):
+
+            def client_connected(self, reader, writer):
+                original_write = writer.write
+
+                def collecting_write(data):
+                    original_write(data)
+                    written.append(data)
+                writer.write = collecting_write
+                super().client_connected(reader, writer)
+
+            @asyncio.coroutine
+            def handshake(self, *args, **kwargs):
+                raise exception_cls()
+
+        self.start_server(klass=WriteCollectingProtocol)
         # Exception appears to be platform-dependent: InvalidHandshake on
         # macOS, ConnectionResetError on Linux. This doesn't matter; this
         # test primarily aims at covering a code path on the server side.
```

Comment:
You can fix the merge conflicts, add some debug code, and look in the Circle CI output to check the behavior on Linux.

This isn't very convenient because every iteration takes two or three minutes, but it's bearable for small checks.

It's late here, I'll investigate the remainder of your comment another day!